### PR TITLE
Fix timing attack

### PR DESCRIPTION
**Pull request by bowenwang1996**
When producing a block, we should not produce timestamp that is older than the previous block. Otherwise a non-malicious block producer can produce an invalid block.

### CODE REVIEW

**Comment by fckt** [comment]
Small question: I see a change in producing logic, could you please point me how the correct block timestamp is validated?

File: chain/chain/tests/simple_chain.rs [hunk 1]
```diff
@@ -178,3 +182,19 @@ fn blocks_at_height() {
...
+    init_test_logger();
+    let (mut chain, _, _, signer) = setup();
+    let genesis = chain.get_block_by_height(0).unwrap();
+    let mut b1 = Block::empty_with_height(genesis, 1, signer.clone());
+    b1.header.inner.timestamp = to_timestamp(b1.header.timestamp() + Duration::seconds(60));
+    let hash = hash(&b1.header.inner.try_to_vec().expect("Failed to serialize"));
+    b1.header.hash = hash;
+    b1.header.signature = signer.sign(hash.as_ref());
+    assert!(chain.process_block(&None, b1.clone(), Provenance::NONE, |_| {}, |_| {}).is_ok());
+    let b2 = Block::empty_with_height(&b1, 2, signer.clone());;
```
Comment (ilblackdragon) [inline chain/chain/tests/simple_chain.rs]: hm, I think you should test this in client:
```
client.process_block(invalid_timestamp_block)
let new_block = client.produce_block().unwrap();
assert!(client.process_block(new_block).is_ok());
```
and pretty much is should have failed before this change.

File: core/primitives/src/block.rs [hunk 2]
```diff
@@ -316,6 +316,12 @@ impl Block {
         let num_approvals: u128 =
             approval_mask.iter().map(|x| if *x { 1u128 } else { 0u128 }).sum();
         let total_weight = prev.inner.total_weight.next(num_approvals);
+        let now = Utc::now();
+        let time = if to_timestamp(now) <= prev.inner.timestamp {
```
Comment (ilblackdragon) [inline core/primitives/src/block.rs]: should we just convert to timstamp here?
e.g. let now = to_timestamp(Utc::now()) / compare or +1 is easier. And pass that to block production.

**Comment by bowenwang1996** [comment]
@fckt the check is in `chain::validate_header`

**Pull request merged**